pkg/controller/lambda/function/setup.go

```diff
@@ -0,0 +1,499 @@
+		return false, nil
+	}
+
+	// Compare CONFIGURATION
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Description), aws.StringValue(obj.Configuration.Description)) != 0 {
+		return false, nil
+	}
+
+	if !isUpToDateEnvironment(cr, obj) {
+		return false, nil
+	}
+
+	// Connection settings for an Amazon EFS file system.
+	if !isUpToDateFileSystemConfigs(cr, obj) {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Handler), aws.StringValue(obj.Configuration.Handler)) != 0 {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.KMSKeyARN), aws.StringValue(obj.Configuration.KMSKeyArn)) != 0 {
+		return false, nil
+	}
+
+	// The function's layers (https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html).
+	if !isUpToDateLayers(cr, obj) {
+		return false, nil
+	}
+
+	// set default
+	if aws.Int64Value(cr.Spec.ForProvider.MemorySize) != 0 &&
+		aws.Int64Value(cr.Spec.ForProvider.MemorySize) != aws.Int64Value(obj.Configuration.MemorySize) {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Role), aws.StringValue(obj.Configuration.Role)) != 0 {
+		return false, nil
+	}
+
+	if strings.Compare(aws.StringValue(cr.Spec.ForProvider.Runtime), aws.StringValue(obj.Configuration.Runtime)) != 0 {
+		return false, nil
+	}
+	// Set default
+	if aws.Int64Value(cr.Spec.ForProvider.Timeout) != 0 &&
+		aws.Int64Value(cr.Spec.ForProvider.Timeout) != aws.Int64Value(obj.Configuration.Timeout) {
+		return false, nil
+	}
+
+	if !isUpToDateTracingConfig(cr, obj) {
+		return false, nil
+	}
+
+	if !isUpToDateSecurityGroupIDs(cr, obj) {
+		return false, nil
+	}
+
+	addTags, removeTags := diffTags(cr.Spec.ForProvider.Tags, obj.Tags)
+	return len(addTags) == 0 && len(removeTags) == 0, nil
+
+}
+
+// isUpToDateEnvironment checks if FunctionConfiguration EnvironmentResponse Variables are up to date
+func isUpToDateEnvironment(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	envVars := map[string]*string{}
+	awsVars := map[string]*string{}
+	if !reflect.ValueOf(cr.Spec.ForProvider.Environment).IsZero() &&
+		!reflect.ValueOf(cr.Spec.ForProvider.Environment.Variables).IsZero() {
+		envVars = cr.Spec.ForProvider.Environment.Variables
+	}
+	if !reflect.ValueOf(obj.Configuration.Environment).IsZero() &&
+		!reflect.ValueOf(obj.Configuration.Environment.Variables).IsZero() {
+		awsVars = obj.Configuration.Environment.Variables
+	}
+	// Check if counts match which should handle nil/empty
+	if len(envVars) != len(awsVars) {
+		return false
+	}
+	// Check if actual values match, and only DeepEqual if there are actual values
+	if len(envVars) > 0 &&
+		!reflect.DeepEqual(envVars, awsVars) {
+		return false
+	}
+	return true
+}
+
+func isUpToDateFileSystemConfigs(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) bool {
+	// Handle nil pointer refs
+	var fileSystemConfigs []*svcsdk.FileSystemConfig
```

Comment:
What do you think about doing a length check first, and then sort both arrays by ARN and then compare if the fields are equal for each pair?

Reply:
What you suggested is the best option.  I have made the change, thanks again.